### PR TITLE
Arf/87799/arp 21a

### PR DESCRIPTION
**Pull request by nihil2501**
Modifications in order (these are just the commits listed out in order):
1. Empty react-router v3 App.
2. Import 21a, overriding intro sign-in. Basic SIP protocol.
3. Support access token refresh & refactor fetchUser.
4. Repair SIP no-auth handling.
5. Comply with log out instruction when SIP notices no-auth.
6. Work around submitted form staleness w/ logo a tag.
7. SIP login links redirect to sign-in.
8. Fix ConfirmationPage submission timestamp proptype.
9. Revert "Repair SIP no-auth handling."
10. Streamline and correct fetch user exceptions.

### Future Extensions
- `no-auth` on submit has incorrect copy for us "Go Back to VA.gov" in `GenericError`
  - Could we customize this?
  - `href` will be fine once we have subdomain
### SIP (Semi-)Bugs
- `no-auth` UX no longer shows up (pursuing platform forms level issue)
- `SaveStatus` shows over-persisted state of previous IPF after starting over
- IPF and form client state persists after submit
- Oddly, all errors on remove API call while starting over become `no-auth`, even when they aren't and UX is weird here

### CODE REVIEW

File: src/applications/accredited-representative-portal/actions/user.js [hunk 1]
```diff
@@ -5,33 +5,42 @@ export const FETCH_USER = 'FETCH_USER';
...
+      const user = await apiRequest(`${environment.API_URL}${path}`);
 
-export const fetchUser = () => async dispatch => {
-  dispatch(fetchUserStart());
+      /**
+       * This is an even stricter success condition than having a user. We
+       * additionally require what is needed for access token refreshing to
+       * function.
+       */
+      const serviceName = user?.profile?.signIn?.serviceName;
```
Comment (gabezurita) [inline src/applications/accredited-representative-portal/actions/user.js:23]: Nice!

File: src/applications/accredited-representative-portal/reducers/user.js [hunk 2]
```diff
@@ -1,38 +1,109 @@
...
-        isLoading: false,
-        profile: null,
-        error: action.error || 'Unknown error',
+        /**
+         * For now, conservatively consider not just `401` but indeed any
+         * momentary inability to retrieve an authenticated user as undeserving
+         * of access to the portions of the app that require authentication.
+         */
+        ...getNullProfileState(),
+        loading: false,
```
Comment (gabezurita) [inline src/applications/accredited-representative-portal/reducers/user.js:59]: Great improvement!

**Review by gabezurita — COMMENTED** [review]
Looking good so far! I'll fuss test this on Monday!

Some good improvements!

File: src/applications/accredited-representative-portal/actions/user.js [hunk 3]
```diff
@@ -5,33 +5,42 @@ export const FETCH_USER = 'FETCH_USER';
...
-  try {
-    // NOTE: accredited_representative_portal/v0/user returns a flat profile object.
-    // see: https://github.com/department-of-veterans-affairs/vets-api/blob/master/modules/accredited_representative_portal/app/controllers/accredited_representative_portal/v0/representative_users_controller.rb
-    // TODO: consider nesting the ARP user profile data in a profile object to match: https://github.com/department-of-veterans-affairs/vets-api/blob/master/app/controllers/v0/users_controller.rb
-    const profile = await requestUserProfile();
-    if (profile) {
-      dispatch(fetchUserSuccess(profile));
-    } else {
-      dispatch(fetchUserFailure('Profile not found'));
+      if (serviceName) {
```
Comment (gabezurita) [inline src/applications/accredited-representative-portal/actions/user.js:25]: If there's no `serviceName`, could we make error handling more explicit?
```js
export function fetchUser() {
  return async (dispatch) => {
    dispatch({ type: FETCH_USER });

    try {
      const path = '/accredited_representative_portal/v0/user';
      const user = await apiRequest(`${environment.API_URL}${path}`);

      if (!user?.profile?.signIn?.serviceName) {
        throw new Error('Service name is missing in the user profile.');
      }

      sessionStorage.setItem('serviceName', user.profile.signIn.serviceName);

      dispatch({
        type: FETCH_USER_SUCCESS,
        payload: user,
      });
    } catch (error) {
      const errorMessage = error.errors?.[0]?.detail || error.message || 'An unknown error occurred while fetching user data.';
      dispatch({
        type: FETCH_USER_FAILURE,
        error: errorMessage,
      });
    }
  };
}
```

**Review by ndsprinkle — APPROVED** [review]

**Review by ndsprinkle — CHANGES_REQUESTED** [review]
Removing approval since you indicated in slack that this is the old/no longer relevant solution?

**Comment by nihil2501** [comment]
> Removing approval since you indicated in slack that this is the old/no longer relevant solution?

@ndsprinkle I reverted the single commit that pertained to that long discussion in Slack you were referring to. The rest of this PR is changes purely within our application. Except for one addition to configure the SIP API URL for our form (our app has its own backend), and for this reason it looks like I still need an approval from forms team for this one line configuration change.

**Comment by tjheffner** [comment]
Verified locally that save-in-progress is working for 21a! 🎉 

vets-website branch: `arf/87799/arp-21a`
vets-api branch: `arf_86023/in_progress_forms.poc`

Signed in via id.me with `vets.gov.user+0@gmail.com`

Screenshot 2024-08-23 at 2 17 47 PM


There are some failing CI checks here (tests need repaired based on some of the changes in the PR), but overall this is awesome work. Thank you @nihil2501 (& @athifw) !

**Comment by gabezurita** [comment]
@nihil2501, I just finished merging `main` and resolving 18 merge conflicts. If/when I get CI green, I'll give this one last round of QA and then approve/merge it!
image

File: src/applications/accredited-representative-portal/reducers/user.js [hunk 4]
```diff
@@ -1,38 +1,109 @@
...
+    lastName: null,
+    verified: false,
+    prefillsAvailable: [],
+    savedForms: [],
+    loading: false,
+  };
+}
 
-const arpUserReducer = (state = initialState, action) => {
+function transformProfilePayload(payload) {
```
Comment (gabezurita) [inline src/applications/accredited-representative-portal/reducers/user.js:23]: Thanks for adding this transform function

File: src/applications/accredited-representative-portal/tests/e2e/accredited-representative-portal.cypress.spec.js [hunk 5]
```diff
@@ -10,32 +10,21 @@ const vamcUser = {
...
-    statusCode: 200,
-    body: {},
-  }).as('getUser');
   setFeatureToggles(featureToggles);
   cy.visit('/representative');
   cy.injectAxe();
 };
 
-describe('Accredited Representative Portal', () => {
+describe.skip('Accredited Representative Portal', () => {
```
Comment (gabezurita) [inline src/applications/accredited-representative-portal/tests/e2e/accredited-representative-portal.cypress.spec.js:27]: This test has stale 21-22 code that needs to be cleaned up.

We'll fix this `.skip` as a part of the below code cleanup issue:
https://app.zenhub.com/workspaces/accredited-representative-facing-team-65453a97a9cc36069a2ad1d6/issues/gh/department-of-veterans-affairs/va.gov-team/92083

**Review by gabezurita — DISMISSED** [review]
LGTM!

**Comment by gabezurita** [comment]
@ndsprinkle, @platform-design-system-fe, @platform-va-product-forms, @department-of-veterans-affairs/va-platform-cop-frontend this is ready for your review!

**Comment by gabezurita** [comment]
@ndsprinkle, @department-of-veterans-affairs/platform-design-system-fe , @department-of-veterans-affairs/platform-va-product-forms, @department-of-veterans-affairs/va-platform-cop-frontend this is ready for your review!

**Review by gabezurita — APPROVED** [review]
Re-approving post mainline merge

**Review by rmessina1010 — APPROVED** [review]
LGTM

**Comment by gabezurita** [comment]
Hey @ndsprinkle, FYI, it seems we need approval from you

**Review by ndsprinkle — APPROVED** [review]
Hey @nihil2501 , looks great to me! Just one thing I wanted to call out, because I'm slightly confused as to what's happening with the manifest.json. Is one manifest being deleted and the app is updated to point to another already existing manifest? Just asking because content-build could break depending on how you did this change.

Here's the information:
```
## Are you removing, renaming or moving a folder in this PR?
- [ ] No, I'm not changing any folders (skip to Summary and delete the rest of this section)
- [ ] Yes, I'm removing, renaming or moving a folder

If the folder you changed contains a `manifest.json`, search for its `entryName` in the content-build [registry.json](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json) (the `entryName` there will match).

If an entry for this folder exists in content-build and you are:
1. **Deleting a folder**:
   1. First search `vets-website` for _all_ instances of the `entryName` in your `manifest.json` and remove them in a separate PR. Look particularly for references in `src/applications/static-pages/static-pages-entry.js` and `src/platform/forms/constants.js`. _**If you do not do this, other applications will break!**_
      - _Add the link to your merged vets-website PR here_
   2. Then, Delete the application entry in [registry.json](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json) and merge that PR **before** this one
      - _Add the link to your merged content-build PR here_

2. **Renaming or moving a folder**: Update the entry in the [registry.json](https://github.com/department-of-veterans-affairs/content-build/blob/main/src/applications/registry.json), but do not merge it until your vets-website changes here are merged. The content-build PR must be merged immediately after your vets-website change is merged in to avoid CI errors with content-build (and Tugboat).
```

**Comment by raudabaugh** [comment]
@ndsprinkle Good callout. The plan is to merge https://github.com/department-of-veterans-affairs/content-build/pull/2180 after this

**Comment by gabezurita** [comment]
We'll first merge the below to be on the safe side:
[devops](https://github.com/department-of-veterans-affairs/devops/pull/14686)
[vsp-platform-revproxy](https://github.com/department-of-veterans-affairs/vsp-platform-revproxy/pull/622)
[content-build](https://github.com/department-of-veterans-affairs/content-build/pull/2269)

**Pull request merged**